src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopy = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'addProperties':
+        Object.assign(stateCopy, extraData);
+        break;
+
+      case 'removeProperties':
+        keysToRemove.forEach(function (key) {
+          delete stateCopy[key];
+        });
+        break;
+
+      case 'clear':
+        for (const key in stateCopy) {
+          if (stateCopy.hasOwnProperty(key)) {
+            delete stateCopy[key];
+          }
+        }
+        break;
+
+      default:
```

Comment:
[CODE STYLE]: Use a default case in your switch statement to handle unexpected action types. It's good practice to throw an error or at least console.warn to inform about an unrecognized action type.